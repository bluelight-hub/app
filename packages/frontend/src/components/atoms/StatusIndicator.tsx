import { Badge, Button, Popover, message } from 'antd';
import React from 'react';
import { PiCopy } from 'react-icons/pi';
import { ConnectionStatus, useBackendHealth } from '../../hooks/useBackendHealth';

/**
 * Props für die StatusIndicator-Komponente
 */
interface StatusIndicatorProps {
    /**
     * CSS-Klasse für Styling-Anpassungen
     */
    className?: string;
    /**
     * Optional callback when status changes
     */
    onStatusChange?: (status: ConnectionStatus) => void;
    /**
     * Whether to show the text
     */
    withText?: boolean;
    /**
     * Test-ID für automatisierte Tests
     */
    "data-testid"?: string;
}

/**
 * Einfache Komponente zur Anzeige des Backend-Verbindungsstatus
 * 
 * Zeigt den aktuellen Verbindungsstatus als farbigen Punkt an und bietet
 * bei Hover detaillierte Informationen zum Systemstatus.
 */
export const StatusIndicator: React.FC<StatusIndicatorProps> = ({
    className,
    onStatusChange,
    withText = false,
    "data-testid": dataTestId = "status-indicator",
}) => {
    // Backend-Gesundheitsstatus vom Hook abrufen
    const {
        getConnectionStatus,
        getSystemDetailsText,
        getDebugInfo
    } = useBackendHealth();

    // Status aktualisieren, wenn sich etwas ändert
    React.useEffect(() => {
        if (onStatusChange) {
            onStatusChange(getConnectionStatus());
        }
    }, [getConnectionStatus, onStatusChange]);

    // Status-Informationen basierend auf dem aktuellen Status erzeugen
    const getStatusInfo = () => {
        const status = getConnectionStatus();
        const systemDetails = getSystemDetailsText();

        switch (status) {
            case 'checking':
                return {
                    badgeStatus: 'processing' as const,
                    text: 'Prüfe Verbindung...',
                    details: 'Verbindungsstatus wird geprüft'
                };
            case 'online':
                return {
                    badgeStatus: 'success' as const,
                    text: 'Vollständig verbunden',
                    details: `Online-Modus (Internet + FüKW)\n${systemDetails}`
                };
            case 'offline':
                return {
                    badgeStatus: 'warning' as const,
                    text: 'Lokaler Modus',
                    details: `Offline-Modus (Nur FüKW)\n${systemDetails}`
                };
            case 'error':
                return {
                    badgeStatus: 'error' as const,
                    text: 'Keine Verbindung',
                    details: `Keine Verbindung zum FüKW\n${systemDetails}`
                };
        }
    };

    const { badgeStatus, text, details } = getStatusInfo();

    // Debug-Informationen in die Zwischenablage kopieren
    const copyToClipboard = async () => {
        const debugInfo = getDebugInfo();
        if (!debugInfo) {
            message.error('Keine Daten zum Kopieren verfügbar');
            return;
        }

        try {
            await navigator.clipboard?.writeText?.(debugInfo);
            message.success('Debug-Informationen in Zwischenablage kopiert');
        } catch {
            message.error('Fehler beim Kopieren der Debug-Informationen');
        }
    };

    // Inhalt des Popovers mit Details
    const popoverContent = (
        <div className="max-w-sm">
            <div className="whitespace-pre-line">
                {text}
                {details && (
                    <>
                        <br />
                        <br />
                        <strong>Systemstatus:</strong>
                        <br />
                        {details}
                    </>
                )}
            </div>
            <div className="mt-3 flex justify-end">
                <Button
                    size="small"
                    icon={<PiCopy className="mr-1" />}
                    onClick={copyToClipboard}
                    title="Debug-Informationen kopieren"
                >
                    Debug-Info kopieren
                </Button>
            </div>
        </div>
    );

    return (
        <Popover
            content={popoverContent}
            title="Verbindungsstatus"
            trigger="hover"
            placement="right"
        >
            <div className="flex items-center gap-x-2">
                <Badge
                    status={badgeStatus}
                    className={className}
                    data-testid={dataTestId}
                />
                <p className="text-sm text-gray-500 dark:text-gray-400">
                    {withText && text}
                </p>
            </div>
        </Popover>
    );
}; 