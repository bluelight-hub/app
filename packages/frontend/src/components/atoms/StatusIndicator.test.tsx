import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { StatusIndicator } from './StatusIndicator';

// Mock des useBackendHealth Hooks
vi.mock('../../hooks/useBackendHealth', () => {
    return {
        useBackendHealth: () => ({
            isLoading: false,
            getConnectionStatus: () => 'online',
            getSystemDetailsText: () => 'internet: ✅\nfuekw: ✅',
            getDebugInfo: () => JSON.stringify({
                status: 'online',
                details: { internet: { status: 'up' }, fuekw: { status: 'up' } },
                timestamp: '2023-01-01T12:00:00.000Z'
            })
        }),
        // Export des ConnectionStatus-Typs als String-Union
        ConnectionStatus: undefined // Dummy, wird nicht genutzt
    };
});

/**
 * Tests für die StatusIndicator-Komponente
 */
describe('StatusIndicator', () => {
    // Unit Test - Komponente rendert ohne Fehler
    it('sollte ohne Fehler rendern', () => {
        render(<StatusIndicator />);
        const indicator = screen.getByTestId('status-indicator');
        expect(indicator).toBeInTheDocument();
    });

    // Unit Test - Klassen werden korrekt angewendet
    it('sollte benutzerdefinierte Klassen akzeptieren', () => {
        render(<StatusIndicator className="custom-class" />);
        const indicator = screen.getByTestId('status-indicator');
        expect(indicator).toHaveClass('custom-class');
    });

    // Unit Test - Test-ID kann überschrieben werden
    it('sollte eine benutzerdefinierte Test-ID akzeptieren', () => {
        render(<StatusIndicator data-testid="custom-indicator" />);
        const indicator = screen.getByTestId('custom-indicator');
        expect(indicator).toBeInTheDocument();
    });

    // Verhalten bei withText=false (Standard)
    it('sollte standardmäßig keinen Text anzeigen', () => {
        render(<StatusIndicator />);
        expect(screen.queryByText('Vollständig verbunden')).not.toBeInTheDocument();
    });

    // Verhalten bei withText=true
    it('sollte Text anzeigen, wenn withText=true', () => {
        render(<StatusIndicator withText={true} />);
        expect(screen.getByText('Vollständig verbunden')).toBeInTheDocument();
    });

    // onStatusChange Callback
    it('sollte den onStatusChange callback aufrufen', () => {
        const onStatusChange = vi.fn();
        render(<StatusIndicator onStatusChange={onStatusChange} />);
        expect(onStatusChange).toHaveBeenCalledWith('online');
    });

    // Snapshot Test
    it('sollte dem Snapshot entsprechen', () => {
        const { container } = render(<StatusIndicator />);
        expect(container).toMatchSnapshot();
    });

    // Snapshot Test mit Text
    it('sollte dem Snapshot mit Text entsprechen', () => {
        const { container } = render(<StatusIndicator withText={true} />);
        expect(container).toMatchSnapshot();
    });
});
