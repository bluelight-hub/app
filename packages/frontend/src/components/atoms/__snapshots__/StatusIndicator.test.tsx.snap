// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatusIndicator > should match snapshot 1`] = `
<div>
  <span
    aria-describedby="test-id"
    class="ant-badge ant-badge-status ant-badge-not-a-wrapper css-dev-only-do-not-override-1v613y0"
    data-testid="status-indicator"
  >
    <span
      class="ant-badge-status-dot ant-badge-status-processing"
    />
  </span>
</div>
`;

exports[`StatusIndicator > sollte dem Snapshot entsprechen 1`] = `
<div>
  <div
    aria-describedby="test-id"
    class="flex items-center gap-x-2"
  >
    <span
      class="ant-badge ant-badge-status ant-badge-not-a-wrapper css-dev-only-do-not-override-1a3rktk"
      data-testid="status-indicator"
    >
      <span
        class="ant-badge-status-dot ant-badge-status-success"
      />
    </span>
    <p
      class="text-sm text-gray-500 dark:text-gray-400"
    />
  </div>
</div>
`;

exports[`StatusIndicator > sollte dem Snapshot mit Text entsprechen 1`] = `
<div>
  <div
    aria-describedby="test-id"
    class="flex items-center gap-x-2"
  >
    <span
      class="ant-badge ant-badge-status ant-badge-not-a-wrapper css-dev-only-do-not-override-1a3rktk"
      data-testid="status-indicator"
    >
      <span
        class="ant-badge-status-dot ant-badge-status-success"
      />
    </span>
    <p
      class="text-sm text-gray-500 dark:text-gray-400"
    >
      Vollst√§ndig verbunden
    </p>
  </div>
</div>
`;
